[RFR] Complete rewrite in ES6

valeriangalliat opened the pull request:
This drops the Themeleon dependency, it only uses the Themeleon abstract interface but without relying on the library.

pascalduez commented on index.js:
```diff
@@ -1,124 +1,2 @@
-
-  return theme.call(this, dest, ctx);
-};
+require('core-js/shim')
```
Is this still needed with 6to5 >=3.0 ?

valeriangalliat replied:
I think so, I was missing at least the global `Promise` otherwise. And [their polyfill](http://6to5.org/docs/usage/polyfill/) is still relyong on `core-js` for this.

pascalduez replied:
http://6to5.org/docs/usage/transformers/#self-contained
http://6to5.org/docs/usage/transformers/#core-js-aliasing

pascalduez commented:
Looks fancy !

FWeinb commented:
Awesome!

valeriangalliat commented:
Glad you like it!

Note I moved `annotations` to `default.json`, because I thought it was making more sense here, though it can now be overriden by the user, while before `annotations` was set *after* merging the `ctx` with `default.json`. Is this a potential problem?

valeriangalliat commented:
Just rebased on top of 262cc24fdff43040fdd12ce79e932a6c64501891.

When I compile the `scss` directory before and after rewrite, I get the exact same output, the sole difference is the `./` being removed from paths because of b73c989 (I removed this feature since we don't need it in a one page website).

valeriangalliat commented:
I'd like to merge this for 2.1. Everyone's okay with this?

Here's a script to compare `develop` and `es6` output on a given project; copy/paste this in your shell in an empty temporary directory.

```sh
npm i sassdoc

git clone --depth 1 --branch develop git://github.com/SassDoc/sassdoc-theme-default.git develop
cd develop
npm i
sed 's/return minify.*/return html;/' index.js > .index.js
mv .index.js index.js
cd -

git clone --depth 1 --branch es6 git://github.com/SassDoc/sassdoc-theme-default.git es6
cd es6
npm i
sed 's/html => minify.*/html => html)/' src/index.js > src/.index.js
mv src/.index.js src/index.js
make dist
cd -

cat << EOF > test
#!/bin/sh -e

mkdir -p in out

if [ -d "\$1" ]; then
    name=\$(basename "\$1")
    dir=\$1
else
    name=\$(basename "\$1" | sed 's/\.git$//')
    dir=in/\$name

    if [ ! -d "\$dir" ]; then
        git clone --depth 1 "\$1" "\$dir"
    fi
fi

t() {
    node_modules/.bin/sassdoc -v -t "./\$1" --dest "out/\$name-\$1" "\$dir"
}

t develop
t es6

git diff "out/\$name-develop" "out/\$name-es6" || :
EOF

chmod +x test
```

Then you can run:

```sh
./test git://github.com/pascalduez/SassyIcons.git
./test git://github.com/pascalduez/SassyFilters.git
./test git://github.com/pascalduez/SassySVG.git
./test git://github.com/HugoGiraudel/SassyIteratorsGenerators
./test git://github.com/HugoGiraudel/SassyJSON
./test git://github.com/HugoGiraudel/SassyMatrix
./test git://github.com/HugoGiraudel/SassyStrings
./test git://github.com/HugoGiraudel/SassyCast
./test git://github.com/HugoGiraudel/SassySort
./test git://github.com/HugoGiraudel/SassyBitwise
./test git://github.com/HugoGiraudel/SassyLogger
./test git://github.com/HugoGiraudel/SassyGradients
```

At each call, it should show a diff between the `develop` and `es6` output. The only changes should be the `./` removal in URLs.

Tested successfully on my side with the projects in the previous code sample.

pascalduez commented:
Wow @valeriangalliat that's conscientious testing, rad !
Okay for merging.

HugoGiraudel commented on default.json:
```diff
@@ -1,5 +1,11 @@
 {
   "display": {
+    "annotations": {
```
Does that mean this is now customisable from a configuration file?

valeriangalliat replied:
Yes, and I'm asking whether it's okay [here](https://github.com/SassDoc/sassdoc-theme-default/pull/58#issuecomment-72336760). :)

HugoGiraudel commented on src/index.js:
```diff
@@ -0,0 +1,53 @@
+import extend  from 'extend';
+import fs from 'fs';
+import fse from 'fs-extra';
+import { minify } from 'html-minifier';
```
Why is this different from other imports?

valeriangalliat replied:
I don't get the question.

HugoGiraudel replied:
`{ minify }`

valeriangalliat replied:
Oh, because `minify` is not the default export value of `html-minifier`.

In `require`-style, it would different in this way: `var minify = require('html-minifier').minify`.

HugoGiraudel commented on src/index.js:
```diff
@@ -0,0 +1,53 @@
+  sassdocExtras.markdown(ctx);
+  sassdocExtras.display(ctx);
+  sassdocExtras.groupName(ctx);
+  sassdocExtras.shortcutIcon(ctx);
```
Could we allow chaining perhaps? Would it make sense?

valeriangalliat replied:
Would be interesting, not sure how yet (because ideally we need to avoid passing the `ctx` parameter everytime). Please open an issue on `sassdoc-extras`.

pascalduez replied:
Good idea.

HugoGiraudel commented on src/swig.js:
```diff
@@ -0,0 +1,63 @@
+const safe = fn =>
+  fn.safe = true && fn;
+
+const isColor = value => {
```
Isn't it bad practice to have a function that possibly returns nothing?

valeriangalliat replied:
In what case no value is returned?

HugoGiraudel replied:
Well, you have no return statement outside of any conditional block.

valeriangalliat replied:
`try` is not conditional.
